discovery/scaleway/baremetal.go

```diff
@@ -0,0 +1,173 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scaleway
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/scaleway/scaleway-sdk-go/api/baremetal/v1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+type baremetalDiscovery struct {
+	*refresh.Discovery
+	Client    *scw.Client
+	Port      int
+	Zone      string
+	Project   string
+	AccessKey string
+	SecretKey string
+	Name      string
+	Tags      []string
+}
+
+const (
+	baremetalLabelPrefix = metaLabelPrefix + "baremetal_"
+
+	baremetalIDLabel         = baremetalLabelPrefix + "id"
+	baremetalIPIsPublicLabel = baremetalLabelPrefix + "ip_is_public"
+	baremetalIPVersionLabel  = baremetalLabelPrefix + "ip_version"
+	baremetalNameLabel       = baremetalLabelPrefix + "name"
+	baremetalOSNameLabel     = baremetalLabelPrefix + "os_name"
+	baremetalOSVersionLabel  = baremetalLabelPrefix + "os_version"
+	baremetalStatusLabel     = baremetalLabelPrefix + "status"
+	baremetalTagsLabel       = baremetalLabelPrefix + "tags"
+	baremetalTypeLabel       = baremetalLabelPrefix + "type"
+	baremetalZoneLabel       = baremetalLabelPrefix + "zone"
+)
+
+func newBaremetalDiscovery(conf *SDConfig) (*baremetalDiscovery, error) {
+	d := &baremetalDiscovery{
+		Port:      conf.Port,
+		Zone:      conf.Zone,
+		Project:   conf.Project,
+		AccessKey: conf.AccessKey,
+		SecretKey: string(conf.SecretKey),
+		Name:      conf.FilterName,
+		Tags:      conf.FilterTags,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "scaleway_sd", false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	profile, err := LoadProfile(conf)
+	if err != nil {
+		return nil, err
+	}
+	d.Client, err = scw.NewClient(
+		scw.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		scw.WithUserAgent(fmt.Sprintf("Prometheus/%s", version.Version)),
+		scw.WithProfile(profile),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up scaleway client: %w", err)
+	}
+
+	return d, nil
+}
+
+func (d *baremetalDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	api := baremetal.NewAPI(d.Client)
+
+	req := &baremetal.ListServersRequest{
+		Zone:      scw.Zone(d.Zone),
+		ProjectID: scw.StringPtr(d.Project),
```

Comment:
You will need to do
```
if d.Project != "" {
		req.Name = scw.StringPtr(d.Project)
	}
```

otherwise you are making the project mandatory.

Reply:
Without this, I get
msg="Unable to refresh
 target groups" err="scaleway-sdk-go: invalid argument(s): project does not respect constraint,  is not a valid UUID."  
 
 with an empty project.

Reply:
Note: this is probably valable for both baremetal and instance